### PR TITLE
pLeftSemigroupalUpdate

### PR DESCRIPTION
**Pull request by fosskers**
Plus some overall project sprucing. Perhaps should have been two PRs.

Should I add more?

This also bumps the version to `0.4.1`.

### CODE REVIEW

File: src/Configuration/Utils.hs [hunk 1]
```diff
@@ -317,18 +316,6 @@ data AppConfiguration a = AppConfiguration
...
--- configuration and exit.
---
-printConfig ∷ Lens' (AppConfiguration a) Bool
-printConfig = lens _printConfig $ \s a → s { _printConfig = a }
-
--- | The 'ConfigFilesConfig' collects all parameters that determine how
--- configuration files are loaded and parsed.
---
-configFilesConfig ∷ Lens' (AppConfiguration a) ConfigFilesConfig
-configFilesConfig = lens _configFilesConfig $ \s a → s { _configFilesConfig = a }
```
Comment (fosskers) [inline src/Configuration/Utils.hs]: These weren't used anywhere, nor is their parent type `AppConfiguration` exported.

File: src/Configuration/Utils/ConfigFile.hs [hunk 2]
```diff
@@ -276,8 +276,7 @@ pConfigFilesConfig = pure id
...
 dropAndUncaml ∷ Int → String → String
-dropAndUncaml i l
-    | length l < i + 1 = l
-    | otherwise = let (h:t) = drop i l
-        in toLower h : concatMap (\x → if isUpper x then "-" ⊕ [toLower x] else [x]) t
-
+dropAndUncaml _ "" = ""
+dropAndUncaml i l = case drop i l of
+    [] -> l
+    (h:t) -> toLower h : concatMap (\x → if isUpper x then "-" ⊕ [toLower x] else [x]) t
```
Comment (fosskers) [inline src/Configuration/Utils/ConfigFile.hs]: There was a pattern match exhaustion warning here.

File: Setup.hs [hunk 3]
```diff
@@ -1,5 +1,5 @@
 -- ------------------------------------------------------ --
--- Copyright © 2015-2018 Lars Kuhtz <lakuhtz@gmail.com>
+-- Copyright © 2015-2019 Lars Kuhtz <lakuhtz@gmail.com>
```
Comment (larskuhtz) [inline Setup.hs]: I am not a copyright expert, but I think we may add a new line for 2019 here with the copyright of the author of this PR. Similar with other files that have code changes.

Comment (larskuhtz) [reply Setup.hs]: I think that meaning would be that the copyright for 2019 only covers the change that got added in that year. Not 100% sure though.

File: src/Configuration/Utils/Internal/ConfigFileReader.hs [hunk 4]
```diff
@@ -208,7 +208,7 @@ loadRemote conf path = do
     url = getConfigFile path
     policy = _cfcHttpsPolicy conf
     doHttp = liftIO $ do
-        request ← (HTTP.parseUrl $ T.unpack url)
+        request ← (HTTP.parseUrlThrow $ T.unpack url)
```
Comment (larskuhtz) [inline src/Configuration/Utils/Internal/ConfigFileReader.hs]: We may have to raise the lower bound on http-client for this. Well, I guess most of the lower bounds are outdated anyways.

File: src/Configuration/Utils/Monoid.hs [hunk 5]
```diff
@@ -99,6 +100,13 @@ instance (FromJSON a, Monoid a) ⇒ FromJSON (LeftMonoidalUpdate a → LeftMonoi
 pLeftMonoidalUpdate ∷ Monoid a ⇒ O.Parser a → MParser a
 pLeftMonoidalUpdate pElement = mappend ∘ mconcat ∘ reverse <$> many pElement
 
+-- | Like `pLeftMonoidalUpdate`, but works for `Semigroup`s instead. Using this
+-- parser requires the input to have at least one copy (say, for flags that can
+-- be passed multiple times).
+--
+pLeftSemigroupalUpdate ∷ Semigroup a ⇒ O.Parser a → MParser a
+pLeftSemigroupalUpdate pElement = (<>) ∘ fold1 ∘ NEL.fromList ∘ reverse <$> some pElement
```
Comment (larskuhtz) [inline src/Configuration/Utils/Monoid.hs]: Should there also be a `pRightSemigroupalUpdate`?

File: src/Configuration/Utils/Internal/ConfigFileReader.hs (hunk 4 above)
Comment (fosskers) [reply src/Configuration/Utils/Internal/ConfigFileReader.hs]: Comes in at `0.4.30`, looks like.

**Pull request merged**